presto-system-access-control/src/main/java/com/facebook/presto/security/KerberosExactMatchAccessControl.java

```diff
@@ -0,0 +1,147 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.security;
+
+import com.facebook.presto.spi.CatalogSchemaName;
+import com.facebook.presto.spi.CatalogSchemaTableName;
+import com.facebook.presto.spi.PrestoException;
+import com.facebook.presto.spi.security.Identity;
+import com.facebook.presto.spi.security.Privilege;
+import com.facebook.presto.spi.security.SystemAccessControl;
+import com.google.common.base.Splitter;
+
+import java.security.Principal;
+import java.util.List;
+
+import static com.facebook.presto.spi.StandardErrorCode.PERMISSION_DENIED;
+
+public class KerberosExactMatchAccessControl
+        implements SystemAccessControl
+{
+    @Override
+    public void checkCanSetUser(Principal principal, String userName)
+    {
+        List<String> parts = Splitter.on("@").splitToList(principal.getName());
```

Comment:
There's a design decision to be made regarding the hostname component, if there is one.

As implemented, this intentionally requires that if the principal is qualified by a hostname, the username must be too. This means that if you have multiple service accounts on different hosts, they each must log in as their own account on Presto. This would be better from an security and auditability standpoint if we get to the point where you can disable specific accounts, or log queries by the account.

I'll do some testing to see if I can get a principal in here that isn't qualified by a realm name based on the defaults set in /etc/krb5.conf and handle it appropriately if that is possible. It seems like a weird design decision to require individual applications to understand a principal name that doesn't have a realm name and understand the nuances of kerberos themselves.